### PR TITLE
YD-573 Prevent update concurrency

### PR DESCRIPTION
**Pull request by Bert-R**
The approach of checking and updating for the preparation actions was wrong. The entity was loaded first in multiple threads and then updated within a lock, without reloading it. This is addressed now by consolidating all preparation actions in a single method doPreparationsAndCheckCanAccessPrivateData that is called by the controllers before any service call (as alternative for the old canAccessPrivateData).

Some integration tests had to be updated as processing of messages now also happens earlier, as part of some other requests (not just fetching messages).

### CODE REVIEW

File: core/src/main/java/nu/yona/server/subscriptions/service/UserService.java [hunk 1]
```diff
@@ -146,9 +146,61 @@ private void onStart()
 	}
 
 	@Transactional
-	public boolean canAccessPrivateData(UUID id)
+	public boolean doPreparationsAndCheckCanAccessPrivateData(UUID id)
 	{
-		return getUserEntityById(id).canAccessPrivateData();
+		if (!getUserEntityById(id).canAccessPrivateData())
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/UserService.java]: Lock should be put before `getUserEntityById`

File: appservice/src/intTest/HibernateStatsTest.md [hunk 2]
```diff
@@ -4,18 +4,18 @@
...
-| GetMessagesSecond| 0| 1| 9| 0| 0| 0| 1| 0| 8| 0| 128| 0| 3| 0| 27| 0| 0| 0| 11| 36| 0| 0| 0| 2| 2| 67| 67
-| GetDayActivityOverviewsFirst| 0| 4| 11| 0| 0| 0| 1| 0| 3| 0| 34| 0| 2| 0| 11| 0| 0| 0| 2| 1| 0| 0| 0| 2| 2| 2| 2
-| GetDayActivityOverviewsSecond| 0| 1| 9| 0| 0| 0| 1| 0| 3| 0| 31| 0| 2| 0| 6| 0| 0| 0| 1| 2| 0| 0| 0| 2| 2| 2| 2
-| GetWeekActivityOverviewsFirst| 0| 6| 26| 0| 0| 0| 1| 0| 3| 0| 93| 0| 2| 0| 13| 0| 0| 0| 2| 1| 0| 0| 0| 2| 2| 2| 2
-| GetWeekActivityOverviewsSecond| 0| 3| 23| 0| 0| 0| 1| 0| 3| 0| 89| 0| 2| 0| 8| 0| 0| 0| 1| 2| 0| 0| 0| 2| 2| 2| 2
-| GetDayActivityOverviewsWithBuddiesFirst| 0| 7| 51| 0| 0| 0| 1| 0| 4| 0| 256| 0| 3| 0| 30| 0| 0| 0| 14| 5| 0| 0| 0| 2| 2| 3| 3
-| GetDayActivityOverviewsWithBuddiesSecond| 0| 5| 49| 0| 0| 0| 1| 0| 3| 0| 255| 0| 3| 0| 27| 0| 0| 0| 14| 3| 0| 0| 0| 2| 2| 3| 3
-| GetDayActivityDetailsLastFridayFirst| 0| 4| 9| 0| 0| 0| 1| 0| 3| 0| 27| 0| 2| 0| 11| 0| 0| 0| 2| 1| 0| 0| 0| 2| 2| 2| 2
-| GetDayActivityDetailsLastFridaySecond| 0| 1| 6| 0| 0| 0| 1| 0| 2| 0| 20| 0| 2| 0| 5| 0| 0| 0| 1| 1| 0| 0| 0| 2| 2| 2| 2
+| GetUserWithPrivateDataFirst| 0| 7| 15| 0| 0| 0| 1| 0| 6| 0| 69| 0| 3| 0| 26| 0| 0| 0| 9| 5| 0| 0| 0| 2| 2| 3| 3
```
Comment (aristotelos) [inline appservice/src/intTest/HibernateStatsTest.md]: The number of queries has increased, is that acceptable?

**Comment by aristotelos** [comment]
Would it be an option to remove the other locking on user updates by now?

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy [hunk 3]
```diff
@@ -710,10 +707,7 @@ class BasicBuddyTest extends AbstractAppServiceIntegrationTest
 		assertResponseStatusOk(response)
 		appService.getBuddies(bob).size() == 0 // Buddy removed for Bob
 		def buddiesRichard = appService.getBuddies(richard)
-		buddiesRichard.size() == 1 // Buddy not yet removed for Richard
-		Buddy buddy = buddiesRichard[0]
-		buddy.sendingStatus == "REQUESTED"
-		buddy.receivingStatus == "REQUESTED"
+		buddiesRichard.size() == 0 // Buddy removed for Richard (processed during request handling)
```
[the remark this answers is not in the document]
Comment (Bert-R) [reply appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: Removed the second one

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy [hunk 4]
```diff
@@ -678,7 +675,7 @@ class BasicBuddyTest extends AbstractAppServiceIntegrationTest
 
 		then:
 		assertResponseStatusOk(response)
-		appService.getBuddies(bob).size() == 1 // Buddy not yet removed for Bob (didn't process the disconnect)
+		appService.getBuddies(bob).size() == 0 // Buddy removed for Bob (processed during request handling)
```
[the remark this answers is not in the document]
Comment (Bert-R) [reply appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: Removed the second one

File: appservice/src/intTest/HibernateStatsTest.md (hunk 2 above)
Comment (Bert-R) [reply appservice/src/intTest/HibernateStatsTest.md]: Not really, but a failure is worse.  I've logged a ticket to gain it back: https://jira.yona.nu/browse/YD-579

**Comment by Bert-R** [comment]
> Would it be an option to remove the other locking on user updates by now?

No, as these could conflict with implicit updates.

File: appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy [hunk 5]
```diff
@@ -678,7 +675,7 @@ class BasicBuddyTest extends AbstractAppServiceIntegrationTest
```
[the remark this answers is not in the document]
Comment (Bert-R) [reply appservice/src/intTest/groovy/nu/yona/server/BasicBuddyTest.groovy]: Fair point. I've removed the tests.

**Pull request merged**